org.eclipse.lsp4e/src/org/eclipse/lsp4e/DocumentContentSynchronizer.java

```diff
@@ -238,29 +240,29 @@ public void documentAboutToBeSaved() {
 		TextDocumentIdentifier identifier = new TextDocumentIdentifier(uri);
 		// Use @link{TextDocumentSaveReason.Manual} as the platform does not give enough information to be accurate
 		WillSaveTextDocumentParams params = new WillSaveTextDocumentParams(identifier, TextDocumentSaveReason.Manual);
-		List<TextEdit> edits = null;
+
 		try {
-			edits = executeOnCurrentVersionAsync(ls -> ls.getTextDocumentService().willSaveWaitUntil(params))
+			List<TextEdit> edits = executeOnCurrentVersionAsync(ls -> ls.getTextDocumentService().willSaveWaitUntil(params))
 				.get(WILL_SAVE_WAIT_UNTIL_TIMEOUT_IN_SECONDS, TimeUnit.SECONDS);
+			try {
+				LSPEclipseUtils.applyEdits(document, edits);
+			} catch (BadLocationException e) {
+				LanguageServerPlugin.logError(e);
+			}
 		} catch (ExecutionException e) {
 			LanguageServerPlugin.logError(e);
 		} catch (TimeoutException e) {
 			Integer timeoutCount = WILL_SAVE_WAIT_UNTIL_TIMEOUT_MAP.compute(uri,
 					(k, v) -> v == null ? 1 : Integer.valueOf(v + 1));
-			String message = "WillSaveWaitUntil timeouted out after " + Integer.valueOf(WILL_SAVE_WAIT_UNTIL_TIMEOUT_IN_SECONDS).toString() +" seconds for " + uri;  //$NON-NLS-1$//$NON-NLS-2$
+			String message = "On-Save action timeouted out after " + Integer.valueOf(WILL_SAVE_WAIT_UNTIL_TIMEOUT_IN_SECONDS).toString() +" seconds for " + uri;  //$NON-NLS-1$//$NON-NLS-2$
 			if (timeoutCount > WILL_SAVE_WAIT_UNTIL_COUNT_THRESHOLD) {
-				message = message + ", it will no longer be called for this document"; //$NON-NLS-1$
+				message = message + ". On-Save actions will no longer be called for this document"; //$NON-NLS-1$
 			}
-			LanguageServerPlugin.logWarning(message, e);
+			ServerMessageHandler.showMessage("On Save Action Timeout", new MessageParams(MessageType.Error, message)); //$NON-NLS-1$
```

Comment:
Shouldn't the title feature a warning sign ⚠ to get some attention?
Also, as those becomes UI elements and not only logging message, those should be internationalized in the Messages class.

Reply:
Since it is an error message, it already shows an error sign. I will move the message to the Messages class

Reply:
done